tools/c7n_azure/tests/azure_common.py

```diff
@@ -12,115 +12,306 @@
+        return s
+
+    @staticmethod
+    def _replace_tenant_id(s):
+        if "graph.windows.net" in s:
+            return re.sub(
+                r"(?P<prefix>(/|%2F)graph.windows.net(/|%2F))"
+                r"[\da-zA-Z]{8}-([\da-zA-Z]{4}-){3}[\da-zA-Z]{12}",
+                r"\g<prefix>" + DEFAULT_TENANT_ID, s)
+        return s
+
+    @staticmethod
+    def _replace_storage_keys(s):
+        # All usages of storage keys have the word "key" somewhere
+        if "key" in s:
+            return re.sub(
+                r"(?P<prefix>=|\"|:)(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==)",
+                r"\g<prefix>" + DEFAULT_STORAGE_KEY, s)
+        return s
+
+    @staticmethod
+    def _json_extension(path):
+        # A simple transformer keeps the native
+        # cassette naming logic in place
+        return path[:-4] + "json"
+
 
 class BaseTest(TestUtils, AzureVCRBaseTest):
     """ Azure base testing class.
     """
+    def __init__(self, *args, **kwargs):
+        super(BaseTest, self).__init__(*args, **kwargs)
+        self._requires_polling = False
 
     def setUp(self):
         super(BaseTest, self).setUp()
         ThreadHelper.disable_multi_threading = True
 
-        # Patch Poller with constructor that always disables polling
-        self.lro_patch = patch.object(msrest.polling.LROPoller, '__init__', BaseTest.lro_test_init)
-        self.lro_patch.start()
+        # We always patch the date so URLs that involve dates match up
+        self._utc_patch = patch.object(actions, 'utcnow', BaseTest.get_test_date)
+        self._utc_patch.start()
+        self.addCleanup(self._utc_patch.stop)
+
+        if not self._requires_polling:
+            # Patch Poller with constructor that always disables polling
+            # This breaks blocking on long running operations (resource creation).
+            self._lro_patch = patch.object(msrest.polling.LROPoller, '__init__', BaseTest.lro_init)
+            self._lro_patch.start()
+            self.addCleanup(self._lro_patch.stop)
+
+        if self.is_playback():
+            if self._requires_polling:
+                # If using polling we need to monkey patch the timeout during playback
+                # or we'll have long sleeps introduced into our test runs
+                Session._old_client = Session.client
+                Session.client = BaseTest.session_client_wrapper
+                self.addCleanup(BaseTest.session_client_cleanup)
 
-    def tearDown(self):
-        super(BaseTest, self).tearDown()
-        self.lro_patch.stop()
+            if constants.ENV_ACCESS_TOKEN in os.environ:
+                self._tenant_patch = patch('c7n_azure.session.Session.get_tenant_id',
+                                           return_value=DEFAULT_TENANT_ID)
+                self._tenant_patch.start()
+                self.addCleanup(self._tenant_patch.stop)
+
+    def get_test_date(self):
+        if self.cassette.responses:
```

Comment:
(minor) code styling suggestion - although we can define the variable in a nested scope and use it outside of that scope, do want to lift the "default" assignment outside?

Reply:
Which variable are you referring to?    Using `self.blah` in the `setUp()` and adding a cleanup to it?  In the case of test classes treating setup as init seems to be the standard.  I don't actually use them anywhere else but the cleanup does behind the scenes I guess.

Reply:
Oh sorry, I meant `header_date` and `test_date`.